PR review: Code to add a SK Connector for external Vector Databases

tawalke opened the pull request:
### Motivation and Context
This pull request/PR review is to add the ability for the Semantic Kernel to persist embeddings/memory to external vector databases like Qdrant. This submission has modifications and additions to allow for integration into the current SK Memory architecture and the subsequent SDK/API of various Vector Databases. The VectorDB Skill/Connectors has significant changes which are likely more than initial estimations, therefore, requesting a preliminary PR. 

Please help reviewers and future users, providing the following information:

1. Why is this change required?
This change is required in order to allow SK developers/users to persist and search for embeddings/memory from a Vector Database like Qdrant. 

2. What problem does it solve?
Adds capabilities of Long Term Memory/Embedding storage to a vector databases from the Semantic Kernel. 

3. What scenario does it contribute to?
Scenario: Long term, scalability memory storage, retrieval, search, and filtering of Embeddings. 

4. If it fixes an open issue, please link to the issue here.
N/A

### Description

This PR currently includes connection for the Qdrant VectorDB only, removing the initial Milvus VectorDB addition and VectorDB client interfaces for consistency across various external vector databases, which will be provided in forthcoming PR. 

**Addition and Modification of Qdrant.Dotnet SDK**
* Removal of VectorRecord and VectorMetaData replacing with VectorRecordData which inheirits from IEmbeddingwithMetadata, IDataStore, IEmbeddingIndex
* Adding FetchVectorsRequest and FetchVectorsResponse classes for calls the Points/Scroll API of Qdrant vector database to get vectors *without* vectorid by collection name only. Initial not in SDK
* Update of QdrantDB constructor for connection to add port so Qdrant API for both REST/gRPC (which have same calls per Qdrant) are supported for performance and binary data
* Adding Points internal class to support various Qdrant Points API calls
* Adding a default vectorsize if not passed in that is default as ADA model
* Update of IVectorDbCollection.cs to support VectorRecordData class and DataEntry<VectorRecordData>
* Adding method to QdrantCollections for GetAllVectorsAsync method
* Update to QdrantCollections.cs to support updated IVectorDBCollections interface changes.
* Changing almost add signatures of Qdrant Methods to return DataEntry<VectorRecordData<float>> instead of VectorRecord
* Changing of SearchVectorsResponse.cs


**Additons to Skills.Memory.VectorDB**
* Adding new namespace: Skills.Memory.VectorDB
* Adding Qdrant VectorDB client for SK, QdrantVectorDB 
* Adding ILongTermMemoryStore interface for VectorDB
* Creating/Adding Qdrant Memory Store class, QdrantMemoryStore.cs. Adding new method for connecting, retrieving collections and embedding from Qdrant DB in cloud. 

These notes will help understanding how your code works. Thanks!

***Note***
* This does build but it has several warnings as the existing SDK code that was unchanged has separate Logger and other functionality that SK now provides that I am requesting review of possible removal.  
* Question about the need or metho for GetCollections to retrieve ALL collections in a external database which could be significant data request as external vector databases store embeddings:  Would like to discuss possible established limit from SK on pull. 


### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [ ] The code builds clean without any errors or warnings 
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

<!-- Thank you for your contribution to the semantic-kernel repo! -->

tawalke committed:
Committing code for SK Connector for VectorDB from private repo for PR review-TEW

tawalke committed:
Committing code for SK Connector for VectorDB from private repo for PR review-TEW

tawalke committed:
Updating full Memory file that was to be copied to forked repo-TEW

tawalke committed:
Updating full Memory file that was to be copied to forked repo-TEW

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
This should likely be added to the *sln file

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/IVectorDb.cs:
```diff
@@ -0,0 +1,12 @@
+
+using System.Threading.Tasks;
+
+namespace Qdrant.DotNet;
```
1) DotNet feels odd -- do we ever expect a different folder under QDrant that isn't DotNet code?
2) My gut says this should be part of Microsoft.SemanticKernel.Skills.Memory.VectorDB.Qdrant namespace

alexchaomander replied:
I agree. It's already under the dotnet src. In other languages, you can imagine this being under the python/src or typescript/src.

shawncal replied:
I think we can drop the "VectorDB" in the path too:
**dotnet/src/SemanticKernel.Skills/Skills.Memory.Qdrant/**

...and matching namespace:
**Microsoft.SemanticKernel.Skills.Memory.Qdrant**

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
```
Should add some UTs and some kernel-syntax-examples

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/ConsoleLogger.cs:
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
I think this should be deleted and updated to use the logger of the Kernel.

shawncal replied:
@lemillermicrosoft do you mean it should take in the logger provided by the kernel? Or is there a public class in the kernel that can be leveraged?

lemillermicrosoft replied:
Take in an ILogger (and likely a user would then supply the logger that is used by the kernel).

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/Verify.cs:
```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
This too, I think should be removed and use Kernel utilities.

awharrison-28 replied:
This is tough. For context, we had originally hoped that the vector DB connectors would be deployed via their own packages. We've been hesitant to tie these connectors too closely with the Kernel. I think Shawn should weigh in here. 

Agree that it would be much cleaner to use the Kernel utilities, unclear if it makes long-term sense though.

lemillermicrosoft replied:
> be deployed via their own packages

Meaning this project may someday not reference Microsoft.SemanticKernel at all? I'd like to learn more if so.

shawncal replied:
SK.Diagnostics.Verify is not public, so would not be accessible to this lib.

shawncal replied:
> > be deployed via their own packages
> 
> Meaning this project may someday not reference Microsoft.SemanticKernel at all? I'd like to learn more if so.

No. the hope was that we could package the Qdrant SDK separately (from this SK skill) or have this skill use a dotnet qdrant client lib if one exists.

To clarify, this qdrant skill will be its own nuget package (same as SK.Skills.Sqlite), and will maintain a dependency on SK.

lemillermicrosoft committed:
Refactor and improve Qdrant DotNet code

Summary: This commit applies various refactoring and formatting changes to the Qdrant DotNet code, which is a wrapper for the Qdrant vector database. The changes include:

- Breaking long lines into multiple lines for better readability and consistency
- Adding or removing blank lines and spaces where needed
- Using expression-bodied properties and constructors, null coalescing operator, string interpolation, var keyword, and discard variable where possible
- Simplifying some boolean expressions and null checks
- Renaming some variables and methods for clarity
- Adding some validation checks and error messages
- Adding missing namespace and XML documentation
- Fixing a bug in the FetchVectorsResponse class, where the Json parsing was not using the System.Text.Json library
- Fixing a bug in the DeleteCollectionAsync method, where the response status code was not checked
- Refactoring the SearchVectorBySimilarityAsync method to take a DataEntry<VectorRecordData<float>> parameter instead of a VectorRecordData<float> parameter, to be consistent with the FindClosestAsync method
- Fixing the indentation in the project file to use four spaces instead of two

lemillermicrosoft committed:
Refactor and improve Qdrant DotNet code

Summary: This commit applies various refactoring and formatting changes to the Qdrant DotNet code, which is a wrapper for the Qdrant vector database. The changes include:

- Breaking long lines into multiple lines for better readability and consistency
- Adding or removing blank lines and spaces where needed
- Using expression-bodied properties and constructors, null coalescing operator, string interpolation, var keyword, and discard variable where possible
- Simplifying some boolean expressions and null checks
- Renaming some variables and methods for clarity
- Adding some validation checks and error messages
- Adding missing namespace and XML documentation
- Fixing a bug in the FetchVectorsResponse class, where the Json parsing was not using the System.Text.Json library
- Fixing a bug in the DeleteCollectionAsync method, where the response status code was not checked
- Refactoring the SearchVectorBySimilarityAsync method to take a DataEntry<VectorRecordData<float>> parameter instead of a VectorRecordData<float> parameter, to be consistent with the FindClosestAsync method
- Fixing the indentation in the project file to use four spaces instead of two

lemillermicrosoft committed:
Update R# commands in VSCode tasks

Summary: This commit updates the arguments and options for the R# commands in the VSCode tasks.json file. It uses the long-form names for clarity, adds verbosity and profile options for the cleanupcode command, and fixes a typo in the output option for the inspectcode command. These changes improve the usability and readability of the R# commands.

lemillermicrosoft committed:
Merge branch 'sk-connector-vectordb' of https://github.com/tawalke/semantic-kernel into sk-connector-vectordb

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/Verify.cs:
```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    // internal static void Contains<T>(IDictionary<string, T> dictionary, string key, string message)
```
If not removed, clean these commented codes up.

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/QdrantDb.cs:
```diff
@@ -0,0 +1,132 @@
+            throw new VectorDbException(VectorDbException.ErrorCodes.CollectionDoesNotExist, nameof(collectionName));
+        }
+
+        if (test.Status != "green")
```
nit: define constant

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/VectorDistanceType.cs:
```diff
@@ -0,0 +1,39 @@
+namespace Qdrant.DotNet;
+
+[JsonConverter(typeof(JsonStringEnumConverter))]
+public enum VectorDistanceType
```
This seem like something not specific to QDrant, should itbe part of Embeddings namespace?

awharrison-28 replied:
This has been an interesting, long-running conversation. I do agree that the basic enumeration of VectorDistanceType belongs with the embeddings namespace.

shawncal replied:
I think this version (these exact values, ordering) are part of the qdrant API. If we graduate them to the kernel and Embeddings namespace, we'll need to make clear they cannot be changed.

FWIW, for now, I think it's ok to maintain a copy of these in the qdrant namespace. We can refactor stuff out later.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/ConsoleLogger.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace Qdrant.DotNet.Internal.Diagnostics;
+
+public static class ConsoleLogger
```
All public methods and objects need to have documentation

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+        this._qdrantclient = new QdrantVectorDb(host, port);
+    }
+
+    public async IAsyncEnumerable<string> GetCollectionsAsync(CancellationToken cancel = default)
```
While I understand why you don't want to implement this, I don't think that returning null is the answer. 

I suggest implementing as the method would be expected to work and then removing later. Alternatively, there could be a flag for the QdrantMemoryStore itself that indicates how it's going to be used. If it's expected to contain 1000s of collections, have this method return empty. If it's going to be used more episodically, then allow this method to operate normally. Just a thought.

tawalke replied:
Okay I'll add this in and put a limit on it and we can discuss more in depth how we want to  go forward as you noted.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+
+        if (this._qdrantdata.ContainsKey(collection) && this._qdrantdata[collection].ContainsKey(key))
+        {
+            vectorresult = this._qdrantdata[collection][key];
```
I'm a little confused as to what's going on here. Are you caching the entire DB locally?

tawalke replied:
As users add, retrieve, delete embeddings, I'm caching them.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+    public async Task<DataEntry<VectorRecordData<float>>> PutAsync(string collection, DataEntry<VectorRecordData<float>> data,
+        CancellationToken cancel = default)
+    {
+        DataEntry<VectorRecordData<float>> vectordata; //= new DataEntry<VectorRecordData<float>>(data.Key, data.Value);
```
nit: remove defunct comment

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+        return Task.CompletedTask;
+    }
+
+    public IAsyncEnumerable<(VectorRecordData<float>, double)> GetNearestMatchesAsync(string collection, Embedding<float> embedding, int limit = 1,
```
Please implement

tawalke replied:
@awharrison-28  This should be there, initially the updated file was mistakenly not copied over from private repo, but 2nd commit should show this implemented in this PR as I copied it over and pushed. Will verify again now.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+        throw new System.NotImplementedException();
+    }
+
+    IAsyncEnumerable<(IEmbeddingWithMetadata<float>, double)> IEmbeddingIndex<float>.GetNearestMatchesAsync(string collection, Embedding<float> embedding,
```
I think this is dead code

tawalke replied:
The updated file that was copied over has this implemented, in 2nd commit. Verifying now it was pushed successfully. 
![image](https://user-images.githubusercontent.com/2299606/223222970-d575bd9c-75ee-4054-a121-57c102df19fc.png)

tawalke replied:
Commit: https://github.com/tawalke/semantic-kernel/commit/dd3d73b5fd8f67ab582f720d5190589c70bc82d3 removes duplicate, dated file copied over mistakenly.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantVectorDb.cs:
```diff
@@ -0,0 +1,209 @@
+/// An implementation of a client for the Qdrant VectorDB. This class is used to 
+/// connect, create, delete, and get embeddings data from a Qdrant VectorDB instance.
+/// </summary>
+public class QdrantVectorDb
```
nit: Rename to QdrantVectorDbClient for better readability

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+    private readonly ConcurrentDictionary<string, IVectorDbCollection> _qdrantcollection = new();
+    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, DataEntry<VectorRecordData<float>>>> _qdrantdata = new();
+    private QdrantVectorDb _qdrantclient;
+    private readonly int _defaultvectorsize = 1536;
```
Might make sense to add a constructor overload providing a default vector size parameter. Default default = 1536, but the user should be able to set it as well if they know what it would be.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,148 @@
+/// <remarks>The Embedding data is saved to a Qdrant Vector database instance specified in the constructor by url and port.
+/// The embedding data persists between subsequent instances and has similarity search capability.
+/// </remarks>
+public class QdrantMemoryStore : ILongTermMemoryStore<float>
```
What is the difference between QdrantMemoryStore and QdrantMemoryClient? I would think that QdrantMemoryClient is the memory store.

awharrison-28 commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
I think this file is duplicated

tawalke replied:
I think it is. Let me commit fresh to make sure there are not 2 files from copying over.

tawalke replied:
Commit: https://github.com/tawalke/semantic-kernel/pull/1/commits/dd3d73b5fd8f67ab582f720d5190589c70bc82d3 removes duplicate, dated file copied over.

tawalke committed:
Removing duplicate QdrantMemoryStore files-TEW

tawalke committed:
Removing duplicate QdrantMemoryStore files-TEW

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/Exception.cs:
```diff
@@ -0,0 +1,22 @@
+
+using System;
+
+namespace Qdrant.DotNet.Internal.Diagnostics;
```
This is a public class, so "internal" namespace might not be what we want to expose.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/Internal/Diagnostics/VectorDbException.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Qdrant.DotNet.Internal.Diagnostics;
+
+public class VectorDbException : Exception<VectorDbException.ErrorCodes>
```
Same comment: public class should not be in an "Internal" namespace. I'd just drop the "Internal" in all instances of `namespace Qdrant.DotNet.Internal.Diagnostics;`, and change them to
`Microsoft.SemanticKernel.Skills.Qdrant.Diagnostics`

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    async IAsyncEnumerable<(IEmbeddingWithMetadata<float>, double)> IEmbeddingIndex<float>.GetNearestMatchesAsync(string collection, Embedding<float> embedding, int limit, double minRelevanceScore)
```
Remove the `IEmbeddingIndex<float>.` prefix, as you just want to implement this interface method normally, not explicitly. Also make sure to add a modifier (public/private/internal)

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+    async IAsyncEnumerable<(IEmbeddingWithMetadata<float>, double)> IEmbeddingIndex<float>.GetNearestMatchesAsync(string collection, Embedding<float> embedding, int limit, double minRelevanceScore)
+    {
+        DataEntry<VectorRecordData<float>> vectordata = new DataEntry<VectorRecordData<float>>()
+                                                            {
```
Spacing is way off here... remove a bunch of the leading whitespace. It should just be tabbed in once from the previous line.

tawalke replied:
Okay will rerun format

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+    {
+        DataEntry<VectorRecordData<float>> vectordata = new DataEntry<VectorRecordData<float>>()
+                                                            {
+                                                                Value = new VectorRecordData<float>(embedding, new Dictionary<string, object>(), new List<string>()),
```
if you have to pass in new, empty collections as constructor args, something's amiss. This class likely needs additional constructor overloads or default arg values.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+                                                                Timestamp = DateTime.Now,
+                                                            };
+        var result = await this._qdrantclient.SearchVectorBySimilarityAsync(collection, vectordata, minRelevanceScore);
+        await foreach (var match in result.ToAsyncEnumerable())
```
Shouldn't call "ToAsyncEnumerable" and then `await foreach` on it. If it's just an IEnumerable, use good old
`foreach(var match in result)`

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+        yield return null;
+    }
+
+    
```
Please run `dotnet format -v d` to automatically fix all these big blocks of empty whitespace, and a bunch of other formatting rules.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/VectorRecord.cs:
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
This doesn't seem like it's qdrant-specific -- it'll likely be needed for all vector store implementations. @awharrison-28 is this significantly different than what you've created already? Opportunity to align?

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/ILongTermMemoryStore.cs:
```diff
@@ -0,0 +1,16 @@
+//namespace Microsoft.SemanticKernel.Connectors.VectorDB.Qdrant;
+namespace Microsoft.SemanticKernel.Skills.Memory.VectorDB;
+
+internal interface ILongTermMemoryStore<TEmbedding> : IDataStore<VectorRecordData<TEmbedding>>, IEmbeddingIndex<TEmbedding>
```
This is nearly identical to IMemoryStore interface, which is public and available in the Kernel. We should use that (there's not much point of using interface abstractions, if each different memory store implementation brings its own interface).

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+                {
+                    if (!this._qdrantdata.ContainsKey(collection))
+                    {
+                        this._qdrantdata.TryAdd(collection, new ConcurrentDictionary<string, DataEntry<VectorRecordData<float>>>());
```
This looks like a completely parallel database. A cache, sure, but how are we managing the cache? When do we purge? What about concurrency issues? Can we guarantee this qdrant db only has one user? I don't think we should be constructing unbounded data caches.

Looks to me like this whole method can be replaced with
```
public Task<DataEntry<VectorRecordData<float>>?> GetAsync(string collection, string key, CancellationToken cancel = default)
{
    return this._qdrantclient.GetVectorDataAsync(collection, key, cancel);
}
```

Please also plumb through the `CancellatioToken` to the qdrant client async methods

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+        bool result; 
+        try
+        {
+            result = this._qdrantclient.DeleteVectorDataAsync(collection, key).Result;
```
No no no. DeleteVectorDataAsync is an async method, so it should be awaited. Please don't use `.Result`, especially in an awaitable method.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+        {
+            throw new VectorDbException($"Failed to remove vector data from Qdrant {ex.Message}");
+        }
+        return Task.CompletedTask;
```
this return statement will not be needed once you await the DeleteVectorDataAsync

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
```
`<GenerateAssemblyInfo>false</GenerateAssemblyInfo>`
This is specified twice, but should not be specified at all if it's to be packaged as nuget. Why is this here?

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+
+    <PropertyGroup>
+        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+        <AssemblyName>Microsoft.SemanticKernel.Skills.Memory.VectorDB</AssemblyName>
```
Update all 3 mentions of `Microsoft.SemanticKernel.Skills.Memory.VectorDB`
to
`Microsoft.SemanticKernel.Skills.Memory.Qdrant`

tawalke replied:
I had an IVectorDBClient interface that was for consistency across the various VectorDB clients like the initial Milvus implementation I was testing. I would like to do something similar to Langchain where there is a base class/interface across the various vectorstore mplementations. I had the namespace for that reason. I can take this out but wanted to share the rationale.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+        <!-- NuGet Package Settings -->
+        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+        <PackageId>Microsoft.SemanticKernel.Skills.Memory.VectorDB</PackageId>
+        <Title>Semantic Kernel - VectorDB Connector</Title>
```
Qdrant Connector

shawncal commented on .vscode/tasks.json:
```diff
@@ -28,6 +28,9 @@
       "problemMatcher": "$msCompile",
       "group": "build"
     },
```
Revert/sync this file

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/DotNet/IVectorDb.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Qdrant.DotNet;
+
+public interface IVectorDb
```
We should certainly evaluate whether these interfaces are needed here in the qdrant connector, or if they belong in the Kernel as part of IMemoryStore or IEmbeddingIndex

timlaverty replied:
I would expect these interfaces to be part of the kernel.  There will be other vector dbs we will need to support.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+                                                            
+                                                                
+        var result = await this._qdrantclient.SearchVectorBySimilarityAsync(collection, vectordata,minRelevanceScore);
+        await foreach (var rd in result.ToAsyncEnumerable())
```
foreach (var rd in result)

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Qdrant/QdrantMemoryStore.cs:
```diff
@@ -0,0 +1,172 @@
+
+    #region private ================================================================================
+    private readonly ConcurrentDictionary<string, IVectorDbCollection> _qdrantcollection = new();
+    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, DataEntry<VectorRecordData<float>>>> _qdrantdata = new();
```
I don't think we need _qdrantdata. It's just makes things more complicated.

shawncal commented on dotnet/src/SemanticKernel.Skills/Skills.Memory.VectorDB/Skills.Memory.VectorDB.csproj:
```diff
@@ -0,0 +1,34 @@
+    <ItemGroup>
+        <PackageReference Include="Microsoft.CSharp" VersionOverride="4.7.0" />
+        <PackageReference Include="Microsoft.Extensions.Logging.Console" />
+        <PackageReference Include="Newtonsoft.Json" VersionOverride="13.0.2" />
```
Package versions must be specified in Directory.Packages.props file.

We'll want to replace Newtonsoft, but don't worry about it for this first version -- I know this sdk uses it for deserializing to `dynamic`, which System.Text.Json doesn't support.

tawalke replied:
Okay will do, the new classes that were added I did use the System.Text.Json to limit Newtonsoft has much as I could.

awharrison-28 commented:
Can QdrantDBStore seamlessly replace VolatileMemoryStore in Example14_Memory and Example15_MemorySkill of KernelSyntaxExamples? 

At my current review, I do not believe that it can. I want to see that before approving this PR.

tawalke committed:
Adding commits from last week mistakenly pushed to SK mirror repo vs. this SK forked repo. Commits for review added for addressing PR Comments-TEW

tawalke committed:
Adding commits from last week mistakenly pushed to SK mirror repo vs. this SK forked repo. Commits for review added for addressing PR Comments-TEW

tawalke committed:
Merging SK main upstream with qdrant connector branch for PR push to SK public repo-TEW

tawalke committed:
Merging SK main upstream with qdrant connector branch for PR push to SK public repo-TEW

tawalke committed:
Adding merged files for Test/AI/OpenAI-TEW

tawalke committed:
Adding merged files for Test/AI/OpenAI-TEW

tawalke committed:
Merge branch 'microsoft:main' into sk-connector-vectordb

lemillermicrosoft committed:
Add QdrantMemoryStore and update VolatileMemoryStore

      Summary: This commit adds a new class QdrantMemoryStore that implements the ILongTermMemoryStore interface using Qdrant as the backend. It also updates the VolatileMemoryStore class to remove a redundant method and fix some async issues. Additionally, it updates the SK-dotnet solution file to include the Skills.Memory.VectorDB project.

lemillermicrosoft committed:
Squashed commit of the following:

commit 3b6156a63127e37e51a58d11b811811f81dd0b9a
Author: Tara E. Walker <taraew@live.com>
Date:   Wed Mar 15 17:56:38 2023 -0400

    Using this branch as repair branch. Cleaning up all files that were deleted from official PR branch and validating all files to push to PR in original-TEW

commit f9068b1930044c1d1bc143dc611b14f17eca4295
Merge: 45e0af1 e7029a6
Author: Tara W <tawalke@users.noreply.github.com>
Date:   Tue Mar 14 16:19:21 2023 -0400

    Merge branch 'microsoft:main' into sk-connector-vectordb

commit 45e0af11478f2a950005a2fc57c03d8e13b6bd47
Author: Tara E. Walker <taraew@live.com>
Date:   Tue Mar 14 06:03:30 2023 -0400

    Adding merged files for Test/AI/OpenAI-TEW

commit fc68934c999b81d9c1671f4ac1ce68b51f80cedc
Author: Tara E. Walker <taraew@live.com>
Date:   Tue Mar 14 06:02:20 2023 -0400

    Merging SK main upstream with qdrant connector branch for PR push to SK public repo-TEW

commit d31554fe356ed5944293a4ea41865e3942952427
Author: Tara E. Walker <taraew@live.com>
Date:   Mon Mar 13 05:46:14 2023 -0400

    Adding commits from last week mistakenly pushed to SK mirror repo vs. this SK forked repo. Commits for review added for addressing PR Comments-TEW

commit dd3d73b5fd8f67ab582f720d5190589c70bc82d3
Author: Tara E. Walker <taraew@live.com>
Date:   Mon Mar 6 16:16:03 2023 -0500

    Removing duplicate QdrantMemoryStore files-TEW

commit af2de3c0c367af9a89e293ace5c5bd267b3552fe
Merge: 2e0a50b 4a594e3
Author: lemiller@microsoft.com <lemiller@microsoft.com>
Date:   Mon Mar 6 11:02:50 2023 -0800

    Merge branch 'sk-connector-vectordb' of https://github.com/tawalke/semantic-kernel into sk-connector-vectordb

commit 2e0a50b58993d72c0e8fbb5d904a2dc0eb25e9af
Author: lemiller@microsoft.com <lemiller@microsoft.com>
Date:   Mon Mar 6 11:01:05 2023 -0800

    Update R# commands in VSCode tasks

    Summary: This commit updates the arguments and options for the R# commands in the VSCode tasks.json file. It uses the long-form names for clarity, adds verbosity and profile options for the cleanupcode command, and fixes a typo in the output option for the inspectcode command. These changes improve the usability and readability of the R# commands.

commit 1cd09c14ce3247ab1e4c2dc4e9b1e83333a0109a
Author: lemiller@microsoft.com <lemiller@microsoft.com>
Date:   Mon Mar 6 11:00:32 2023 -0800

    Refactor and improve Qdrant DotNet code

    Summary: This commit applies various refactoring and formatting changes to the Qdrant DotNet code, which is a wrapper for the Qdrant vector database. The changes include:

    - Breaking long lines into multiple lines for better readability and consistency
    - Adding or removing blank lines and spaces where needed
    - Using expression-bodied properties and constructors, null coalescing operator, string interpolation, var keyword, and discard variable where possible
    - Simplifying some boolean expressions and null checks
    - Renaming some variables and methods for clarity
    - Adding some validation checks and error messages
    - Adding missing namespace and XML documentation
    - Fixing a bug in the FetchVectorsResponse class, where the Json parsing was not using the System.Text.Json library
    - Fixing a bug in the DeleteCollectionAsync method, where the response status code was not checked
    - Refactoring the SearchVectorBySimilarityAsync method to take a DataEntry<VectorRecordData<float>> parameter instead of a VectorRecordData<float> parameter, to be consistent with the FindClosestAsync method
    - Fixing the indentation in the project file to use four spaces instead of two

commit 4a594e305bfdccef0f74a5ca8d605b66be0fc27a
Author: Tara E. Walker <taraew@live.com>
Date:   Mon Mar 6 13:47:35 2023 -0500

    Updating full Memory file that was to be copied to forked repo-TEW

commit 962c43702dcc9c399a550910237972d3eaa5c8ae
Author: Tara E. Walker <taraew@live.com>
Date:   Mon Mar 6 07:40:40 2023 -0500

    Committing code for SK Connector for VectorDB from private repo for PR review-TEW

lemillermicrosoft committed:
update vectordb to qdrant project + build+warnings

lemillermicrosoft committed:
formatting

lemillermicrosoft committed:
revert volatilememorystore

lemillermicrosoft committed:
revert test changes

tawalke committed:
Merge branch 'sk-connector-vectordb' of https://github.com/tawalke/semantic-kernel into sk-connector-vectordb

tawalke committed:
Updating files per new feekback request for HTTPSchema namespace change-TEW

tawalke committed:
Revert "Updating files per new feekback request for HTTPSchema namespace change-TEW"

This reverts commit 9ec0567b8791f1aa59cfe62ad572a2bca6a5c93a.

lemillermicrosoft committed:
Fix formatting and merge conflict in Qdrant and VolatileMemoryStore

      Summary: This commit fixes a minor formatting issue in QdrantHttpClientWrapper.cs, where the port assignment was indented too much. It also resolves a merge conflict in VolatileMemoryStore.cs, where the sk-connector-vectordb branch had added a generic type parameter to the class. The conflict was resolved by keeping the generic version and removing the leftover markers.

lemillermicrosoft committed:
Remove vectordb

lemillermicrosoft committed:
HttpSchema rename

tawalke committed:
Updated csproj for Assembly name change to Qdrant, pushed another missing Qdrant API file; updated namespace to HTTPSchema per feedback; Testing no merge issues after rebase-TEW

tawalke committed:
Updating Qdrant files removing formatting warnings-TEW

tawalke committed:
WIP: Cleaning up Qdrant connector HTTP Requests-TEW

tawalke committed:
Updating DataModels for HTTP cleanup-TEW

tawalke committed:
WIP: Added all Qdrant API data models, added constants for relevant Qdrant v1 API constants, Updating HTTP request w/HTTPClient Jsoncontent extension method-TEW

tawalke committed:
Adding files from AH Memory Update, WIP: Cleaning up/Simplifying HTTP Qdrant requests-TEW

tawalke committed:
WIP: Updating Qdrant Http Request Class; Removing Request/Response files no longer needed-TEW

tawalke committed:
Updating HttpRequest class-TEW

tawalke committed:
WIP: Creating Handlers for Collections API-TEW

tawalke committed:
WIP: Update to QdrantResponse for Handlers to return correct Qdrant DataModel-TEW

tawalke committed:
WIP: Update to Collection Data Models-TEW

tawalke committed:
WIP: Completed Create Collection/Testing results-TEW

tawalke committed:
Typo fix to CollectionHandler-TEW

tawalke committed:
Update to namespace issues, correction to CollectionHandler dup call-TEW

tawalke committed:
WIP: Updating CollectionHandler, Removing all the old Response/Request files-TEW

tawalke committed:
WIP: Adding base for Points and Search API Handlers classes-TEW

tawalke committed:
Upate to API Handlers for Points and Search-TEW

tawalke committed:
WIP: Testing CreateCollection, removing the Request/Response calls-TEW

tawalke committed:
Updating files based upon Qdrant Http test w/kernel-syntax samples-TEW

tawalke committed:
Updating QdrantDBClient, HTTP Request, and CollectionHandler

tawalke committed:
WIP: Updating Qdrant Client to reflect changes in API Handler & HTTP Requests. All refs to non-.NET system based json serializers are removed-TEW

tawalke committed:
Update Qdrant DB Client to get collection info API-TEW

tawalke committed:
Updating Http handler and Testing-TEW

tawalke committed:
Updating Http Handler-TEW

tawalke committed:
Updating Collection Handler for HTTP requests-TEW

tawalke committed:
Correcting errors in HTTP Request/Switch-TEW

tawalke committed:
Update switch to test against uppercase, removed exception unused warnings-TEW

tawalke committed:
Work in progress: Removed duplicates - Replace with one file for clarity. Adding/Copying functionality back to that file-TEW

tawalke committed:
Work in progress: Update to new file removing duplicates-fix typo-TEW

tawalke committed:
Removed all duplicates, added all code back to new Points Handler file, Updated DataModel to reflect changes. WIP: Update QdrantClient w/new object-TEW

tawalke committed:
WIP: Cleaning up warnings-TEW

lemillermicrosoft committed:
#r and dotnet format

tawalke committed:
Merge pull request #2 from tawalke/sk-qdrant-skill-httpcleanup

Cleaned up all of the JSON library migration to System.Text; Added Qdrant based Data Models reflective of Qdrant v1.1 from Sk qdrant skill httpcleanup branch

tawalke commented:
Several commits and vast changes made from testing on last week that requesting review from PR reviewers. Testing continuing w/Lee

tawalke committed:
Merging format and local changes-TEW

tawalke committed:
WIP: Qdrant VectorDB Client updates-TEW

tawalke committed:
Merge pull request #3 from tawalke/sk-qdrant-skill-httpcleanup

Sk qdrant skill httpcleanup

shawncal committed:
Merge branch 'memory-qdrant' into sk-connector-vectordb
